File: apps/web/lib/exercise-image.ts
// Pure image-source resolution and client-side upload validation for the curator-only
// Exercise Image (issue #504, ADR-0041). No server-only or React imports, so it is safe in
// both Server and Client Components and unit-testable with `node --test`.
//
// Two responsibilities: decide which picture the detail/editor shows (uploaded-first, then the
// legacy curated URL, then nothing), and pre-check an upload's type/size so the editor can flag
// an obviously-bad file before the round trip. The backend re-validates and stays the real gate.

// The safe raster types a browser renders without a plugin and that carry no active content —
// the same allow-list the backend `validate_image` enforces (SVG is excluded: it can carry
// script). Kept in lockstep with `app/domain/exercise_image.py`.
export const ALLOWED_IMAGE_TYPES = [
  "image/jpeg",
  "image/png",
  "image/webp",
] as const;

// 2 MB — the same ceiling the backend enforces. Generous for a photo, small enough that one
// upload never bloats the app database.
export const MAX_IMAGE_BYTES = 2 * 1024 * 1024;

// The structural subset of the Exercise detail the image source is resolved from: the id (to
// build the served URL), whether an uploaded image exists, and the legacy curated URL.
export interface ExerciseImageSource {
  id: number;
  has_image: boolean;
  image: string | null;
}

// The `<img>` src the Exercise Image should point at, or `null` when the movement has no
// picture (the caller then renders nothing — never a broken image). The uploaded image is
// preferred: it is served same-origin by `GET /api/exercises/{id}/image` (proxied to the API
// with the caller's auth). Otherwise the legacy curated URL is used, and a blank legacy value
// counts as no image.
export function resolveExerciseImageSrc(exercise: ExerciseImageSource): string | null {
  if (exercise.has_image) return `/api/exercises/${exercise.id}/image`;
  const legacy = exercise.image?.trim();
  return legacy ? legacy : null;
}

// The bare media type of an upload, lowercased and stripped of any `;`-parameters — the browser
// may send `image/jpeg; charset=binary` or vary the casing, and the media type alone decides.
function normalizeType(type: string): string {
  return type.split(";", 1)[0]?.trim().toLowerCase() ?? "";
}

// Validate a selected file's type and size before upload. Returns a user-facing message for the
// first problem (type before size, mirroring the backend's 415-before-413), or `null` when the
// file is acceptable.
export function validateImageUpload(file: { type: string; size: number }): string | null {
  const type = normalizeType(file.type);
  if (!(ALLOWED_IMAGE_TYPES as readonly string[]).includes(type)) {
    return "Choose a JPEG, PNG, or WebP image.";
  }
  if (file.size > MAX_IMAGE_BYTES) {
    return "Image must be 2 MB or smaller.";
  }
  return null;
}

File: apps/api/app/domain/exercise_image.py
"""Curator-only Exercise Image validation (issue #504, ADR-0041).

The Exercise Image is the one illustration a curator uploads for a movement — the picture the
Enrichment AI is *forbidden* to fabricate, because a misleading generated image is a safety
hazard in an injury/rehab-cautious domain. An upload is admitted only when it is a real image
of a safe type and a sane size, so this module owns those two rules as pure, unit-tested
predicates: the route reads the verdict and maps it to a status (415 unsupported type, 413 too
large), never re-deciding the policy itself."""

from __future__ import annotations

from enum import Enum

# The safe raster types a browser can render without a plugin and that carry no active
# content: JPEG, PNG, WebP. SVG is deliberately excluded — it is XML that can carry script,
# so it is an XSS vector, not a photo of a movement.
ALLOWED_IMAGE_CONTENT_TYPES: frozenset[str] = frozenset(
    {"image/jpeg", "image/png", "image/webp"}
)

# A sane ceiling for a single illustration: 2 MB is generous for a photo yet small enough
# that one upload never bloats the app database (images live in Postgres, ADR-choice Q14a).
MAX_IMAGE_BYTES: int = 2 * 1024 * 1024


class ImageRejection(str, Enum):
    """Why an image upload is refused — each maps to one HTTP status at the route boundary.

    ``UNSUPPORTED_TYPE`` → 415 (the content-type is not in the allow-list); ``TOO_LARGE`` →
    413 (the file exceeds ``MAX_IMAGE_BYTES``)."""

    UNSUPPORTED_TYPE = "unsupported_type"
    TOO_LARGE = "too_large"


def normalize_content_type(content_type: str | None) -> str:
    """Reduce a raw ``Content-Type`` header to its bare media type, lowercased.

    A browser may send parameters (``image/jpeg; charset=binary``) or vary the casing; the
    media type alone decides the allow-list, so strip any ``;``-parameters and casing first.
    Public so the upload route stores the same normalized type it was validated against,
    rather than re-deriving it inline."""

    if not content_type:
        return ""
    return content_type.split(";", 1)[0].strip().lower()


def validate_image(content_type: str | None, byte_size: int) -> ImageRejection | None:
    """Judge one upload's type and size, returning the rejection reason or ``None`` if allowed.

    Type is checked before size on purpose: a wrong-type upload is refused as ``UNSUPPORTED_TYPE``
    regardless of how large it is, so the answer is deterministic when a file is both. The size
    ceiling is inclusive — a file of exactly ``MAX_IMAGE_BYTES`` is accepted; one byte more is
    ``TOO_LARGE``. Pure: no I/O, so it is trivially unit-testable and the route stays a thin
    adapter that only maps the verdict to an HTTP status."""

    if normalize_content_type(content_type) not in ALLOWED_IMAGE_CONTENT_TYPES:
        return ImageRejection.UNSUPPORTED_TYPE
    if byte_size > MAX_IMAGE_BYTES:
        return ImageRejection.TOO_LARGE
    return None


__all__ = [
    "ALLOWED_IMAGE_CONTENT_TYPES",
    "MAX_IMAGE_BYTES",
    "ImageRejection",
    "normalize_content_type",
    "validate_image",
]
